File: original.cpp
#include "libs.hpp"


//  ----------------------------------------------------------------------
//  Screen handling

LiquidCrystal lcd(8, 9, 4, 5, 6, 7);

int menuPage = 0;
std::string menuItems[] = {"START CAPTURE", "START SHOWCASE" ,"PRESETS",
                           "SET TRIGGER", "SETTINGS", "ABOUT"};
int maxMenuPages = round(((sizeof(menuItems) / sizeof(std::string)) / 2) + .5);

// Bitmap icons drawn on screen
std::byte downArrow[8] = { (std::byte)0b00100, (std::byte)0b00100,
                           (std::byte)0b00100, (std::byte)0b00100,
                           (std::byte)0b00100, (std::byte)0b11111,
                           (std::byte)0b01110, (std::byte)0b00100 };
std::byte upArrow[8]   = { (std::byte)0b00100, (std::byte)0b01110,
                           (std::byte)0b11111, (std::byte)0b00100,
                           (std::byte)0b00100, (std::byte)0b00100,
                           (std::byte)0b00100, (std::byte)0b00100 };

// This function will generate the 2 menu items that can fit on the screen.
// They will change as you scroll through your menu.
// Up and down arrows will indicate your current menu position.
void mainMenuDraw() {
    lcd.clear();
    lcd.setCursor(1, 0);
    lcd.print(menuItems[menuPage]);
    lcd.setCursor(1, 1);
    lcd.print(menuItems[menuPage + 1]);
    if (menuPage == 0) {
        lcd.setCursor(15, 1);
        lcd.write(std::byte(2));
    } else if (menuPage > 0 and menuPage < maxMenuPages) {
        lcd.setCursor(15, 1);
        lcd.write(std::byte(2));
        lcd.setCursor(15, 0);
        lcd.write(std::byte(1));
    } else if (menuPage == maxMenuPages) {
        lcd.setCursor(15, 0);
        lcd.write(std::byte(1));
    }
}


//  ----------------------------------------------------------------------
//  Reacting to button presses

// This function is called whenever a button press is evaluated.
// The LCD shield works by observing a voltage drop across the buttons all
// hooked up to A0.
int evaluateButton(int x) {
    int result = 0;
    if (x < 50) {
        result = 1;  // up
    } else if (x < 195) {
        result = 2;  // down
    }
    return result;
}

int readKey;

void operateMainMenu () {
    int activeButton = 0;
    while (activeButton == 0) {
        int button;
        readKey = analogRead(0);
        button = evaluateButton(readKey);
        switch (button) {
        case 0: // When button returns as 0 there is no action taken
            break;
        case 1:
            button = 0;
            menuPage = menuPage - 1;
            menuPage = constrain(menuPage, 0, maxMenuPages);
            mainMenuDraw ();
            activeButton = 1;
            break;
        case 2:
            button = 0;
            menuPage = menuPage + 1;
            menuPage = constrain(menuPage, 0, maxMenuPages);
            mainMenuDraw ();
            activeButton = 1;
            break;
        }
    }
}


//  ----------------------------------------------------------------------
//  Top level 

void setup() {
    lcd.begin(16, 2);
    lcd.createChar(std::byte(1), upArrow);
    lcd.createChar(std::byte(2), downArrow);
}

void loop() {
    mainMenuDraw();
    operateMainMenu();
}

// To simulate the rest of the toolchain
int main () {
    setup ();
    while (true) {
        loop ();
    }
}

File: ruby/original.rb
# Ruby

require_relative 'libs'

# Screen handling
lcd = Libs::LiquidCrystal.new(8, 9, 4, 5, 6, 7)

menu_page = 0
menu_items = ["START CAPTURE", "START SHOWCASE", "PRESETS", "SET TRIGGER", "SETTINGS", "ABOUT"]
max_menu_pages = ((menu_items.size / 2.0).round) + 0.5

# Bitmap icons drawn on screen
down_arrow = [0b00100, 0b00100, 0b00100, 0b00100, 0b00100, 0b11111, 0b01110, 0b00100]
up_arrow = [0b00100, 0b01110, 0b11111, 0b00100, 0b00100, 0b00100, 0b00100, 0b00100]

def main_menu_draw
  lcd.clear
  lcd.set_cursor(1, 0)
  lcd.print(menu_items[menu_page])
  lcd.set_cursor(1, 1)
  lcd.print(menu_items[menu_page + 1])
  if menu_page == 0
    lcd.set_cursor(15, 1)
    lcd.write(2)
  elsif menu_page > 0 and menu_page < max_menu_pages
    lcd.set_cursor(15, 1)
    lcd.write(2)
    lcd.set_cursor(15, 0)
    lcd.write(1)
  elsif menu_page == max_menu_pages
    lcd.set_cursor(15, 0)
    lcd.write(1)
  end
end

def evaluate_button(x)
  result = 0
  if x < 50
    result = 1  # up
  elsif x < 195
    result = 2  # down
  end
  result
end

def operate_main_menu
  active_button = 0
  while active_button == 0
    read_key = analog_read(0)
    button = evaluate_button(read_key)
    case button
    when 0
      # When button returns as 0 there is no action taken
    when 1
      button = 0
      menu_page -= 1
      menu_page = [menu_page, 0].max
      main_menu_draw
      active_button = 1
    when 2
      button = 0
      menu_page += 1
      menu_page = [menu_page, max_menu_pages].min
      main_menu_draw
      active_button = 1
    end
  end
end

def setup
  lcd.begin(16, 2)
  lcd.create_char(1, up_arrow)
  lcd.create_char(2, down_arrow)
end

def cycle
  main_menu_draw
  operate_main_menu
end

def main
  setup
  loop do
    cycle
  end
end

# Call the main function
main
